engine/class_modules/priest/sc_priest_shadow.cpp

```diff
@@ -74,21 +74,45 @@ struct mind_blast_t final : public priest_spell_t
 
   bool ready() override
   {
-    if ( only_cwc )
+    // Check conditions that do NOT pertain to the target before cycle_targets
+    if ( cooldown->is_ready() == false && !priest().buffs.dark_thoughts->check() )
+      return false;
+
+    if ( internal_cooldown->down() && !priest().buffs.dark_thoughts->check() )
+      return false;
+
+    if ( only_cwc && (!priest().buffs.dark_thoughts->check() || player->channeling == nullptr || player->channeling->data().id() != mind_flay_spell->id() &&
+         player->channeling->data().id() != mind_sear_spell->id() ))
+      return false;
+
+    if ( player->is_moving() && !usable_moving() )
+      return false;
+
+    auto resource = current_resource();
+    if ( resource != RESOURCE_NONE && !player->resource_available( resource, cost() ) )
     {
-      if ( !priest().buffs.dark_thoughts->check() )
-        return false;
-      if ( player->channeling == nullptr )
-        return false;
-      if ( player->channeling->data().id() == mind_flay_spell->id() ||
-           player->channeling->data().id() == mind_sear_spell->id() )
-        return priest_spell_t::ready();
+      if ( starved_proc )
+        starved_proc->occur();
       return false;
     }
-    else
+
+    if ( usable_while_casting )
     {
-      return priest_spell_t::ready();
+      if ( execute_time() > timespan_t::zero() )
+      {
+        return false;
+      }
+
+      // Don't allow cast-while-casting spells that trigger the GCD to be ready if the GCD is still
+      // ongoing (during the cast)
+      if ( ( player->executing || player->channeling ) && gcd() > timespan_t::zero() &&
+           player->gcd_ready > sim->current_time() )
+      {
+        return false;
+      }
     }
+
+    return true;
```

Comment:
what is the benefit of not just using `return priest_spell_t::ready();` here? Dunno what @scamille thinks but I feel like we should have that

Reply:
Yeah we need ready here for this: https://github.com/simulationcraft/simc/blob/shadowlands/engine/class_modules/priest/sc_priest.hpp#L1383-L1386

Reply:
return priest_spell_t::ready() will fail if the cooldown of mindblast is up but you have a dark thoughts proc. I'll manually add https://github.com/simulationcraft/simc/blob/shadowlands/engine/class_modules/priest/sc_priest.hpp#L1383-L1386 so it's compliant.
Both cooldown and internal_cooldown need to be checked because they are part of the base action_t::ready function. Internal cooldown is probably irrelevant though. If they don't get checked then the spell can just be cast since I never call priest_spell_t:ready()

Reply:
hm okay, i see why you did this now, i wish there was a better way :/ @lgkern @scamille any ideas?